src/algorithms/graphs/breadthFirstSearch.java

```diff
@@ -0,0 +1,133 @@
+package src.algorithms.graphs;
+
+import java.util.*;
+
+import src.algorithms.graphs.util.BinaryTreeNode;
+import src.algorithms.graphs.util.GraphNode;
+
+/**
+ * Implementation of BFS
+ *
+ * Breadth-First search is a graph traversal algorithm that utilizes a queue (FIFO) data structure
+ * It is useful in finding a shortest-hops solution in graphs
+ * This method is also used to obtain level order traversals of trees.
+ *
+ * In general, BFS works as such:
+ *  - Start with a queue that contains the root node
+ *  - While the queue is not empty:
+ *      - Pop a vertex from the queue
+ *      - Push all neighbours to the queue if they have not been visited yet
+ *      - Update any variables as needed
+ *
+ * Time: O(V + E), where V is the number of vertices/nodes, and E is the number of edges in the graph
+ * Explanation: Each vertex is popped in O(1) time from the stack exactly once, hence O(V)
+ *              For each edge, we must check in O(1) time if we have visited the adjacent vertex to know
+ *              whether to push it into the queue, hence O(E).
+ *              Note that if the graph is a forest, BFS will likely terminate earlier, as fewer vertices are traversed
+ *
+ * Space: O(V): We utilize a Hashset to store the vertices we have visited already. In the worst case we have a
+ *              connected graph where all vertices are traversed, and our Hashset stores all O(V) vertices.
+ *              Further, we use a queue to hold the vertices to be traversed. In the worst case, the root will have all
```

Comment:
Okay, but this isn't the only worst case. You can just mention the queue can possibly hold up to v-1 nodes, for instance when the root is connected to all other v-1 nodes.